Feature: extend leaf face neighbor to non-balanced forests and ghosts [N/N]

holke opened the pull request:
Closes #1737 and closes #1299

Will also close #328 

**Disclaimer**

~~This PR is not 2k lines as it shows currently.
It is based on many changes that are parts of different PRs but until these are not merged, their lines will show up in the count here.
I listed all the PRs below and also marked the sections that do not need review in the "Files Changed" tab.~~

**_Describe your changes here:_**

~~This has gotten quite large and will be split up and take some time.
I am putting this up here now to start the review early.~~

~~I would like to have a review on this because the algorithm is now working.
The review can focus mostly on the leaf_face_neighbor functionality.
It is still in draft status, because~~

~~a) there is still low-level functionality missing to make this work on all element shapes
b) i will split this up into multiple PRs eventually
c) not all tests pass, due to missing low-level functionality~~

Here is an overview over the exsiting PRs that cover changes from this one.
So a review of this PR can exclude all the listed once below:

- [x] Merge #1301 first (t8_element_array_find, t8_forest_element_is_leaf_or_ghost, t8_forest_tree_is_ghost, t8_forest_element_is_ghost)
- [x] Merge #1714 first (element_is_ancestor and ancestor search, adapt callbacks in test environment)
- [x] Merge #1806 first (t8_cmesh_get_tree_face_neighbor_eclass )
- [x] Merge #1952 first
- [x] Merge #1954 first (Bugfix in VTK empty partitions)
- [x] Merge #1955 first (eclass invalid string)
- [x] Merge #1958 first (element_face_get_ancestor_face)
- [x] Merge #1968 first (const params for face iterator)

**Description**

This PR extends the leaf face neighbor computation in two ways simultaneously:
a) extend it to ghosts, thus we can compute neighbors of ghosts and neighbors that are ghosts
b) extend it to non-balanced forests. So far a 2:1 condition was required, this is no longer the case.

The main contributions are
- extending many "low-level" element and element array handling functions to ghosts (Could be own PR)
- extending the face iteration function to ghosts 
- t8_element_array_find function to search for an element in an array (Should be its own PR)
- t8_forest_leaf_face_neighbors_ext extension to non-balances and ghosts
- t8_forest_same_level_leaf_face_neighbor_index  to compute the data indices of neighbors if only single neighbor (TODO: this was
implemented while the lfn function did not work for adaptive forests, need to check whether to extend this function to all forests now) (Should be its own PR)
- a test for the t8_forest_leaf_face_neighbors_ext and t8_forest_same_level_leaf_face_neighbor_index 
- Common header for unit test adapt callbacks that we can reuse (included in #1714)
- A new functionality to compute the "Ancestor face" - given a face of an element and a coarser level where the face is a subface of, compute the face number at the coarser level

Edit: ~~The function was so far only tested for quads and hexes due to the missing face ancesor #1738. A first test with lines does fail - i need to investigate what is going on. (t8_gtest_face_neighbors/forest_face_neighbors.test_face_neighbors/t8_cmesh_new_from_class_Line_sc_MPI_COMM_WORLD_default)~~
Edit2: It now works for all element shapes :)


What is still open to get this out of draft:
- [x]  Split into multiple PRs
- [x] Implement a scheme function that computes a face index of an ancestor. I.e. Given an element E, a face f of E and and ancestor A (or a level) of E that shares the face, return the face index f_A of the ancestor face of f.
(see #1958)

**_All these boxes must be checked by the AUTHOR before requesting review:_**
- [x] The PR is *small enough* to be reviewed easily. If not, consider splitting up the changes in multiple PRs.
- [x] The title starts with one of the following prefixes: `Documentation:`, `Bugfix:`, `Feature:`, `Improvement:` or `Other:`.
- [x] If the PR is related to an issue, make sure to link it.
- [x] The author made sure that, as a reviewer, he/she would check all boxes below.

**_All these boxes must be checked by the REVIEWERS before merging the pull request:_**

As a reviewer please read through all the code lines and make sure that the code is *fully understood, bug free, well-documented and well-structured*.
#### General
- [ ] The reviewer executed the new code features at least once and checked the results manually.
- [ ] The code follows the [t8code coding guidelines](https://github.com/DLR-AMR/t8code/wiki/Coding-Guideline).
- [ ] New source/header files are properly added to the CMake files.
- [ ] The code is well documented. In particular, all function declarations, structs/classes and their members have a proper doxygen documentation.
- [ ] All new algorithms and data structures are sufficiently optimal in terms of memory and runtime (If this should be merged, but there is still potential for optimization, create a new issue).
#### Tests
- [ ] The code is covered in an existing or new test case using Google Test.
- [ ] The code coverage of the project (reported in the CI) should not decrease. If coverage is decreased, make sure that this is reasonable and acceptable.
- [ ] Valgrind doesn't find any bugs in the new code. [This script](https://github.com/DLR-AMR/t8code/blob/main/scripts/check_valgrind.sh) can be used to check for errors; see also this [wiki article](https://github.com/DLR-AMR/t8code/wiki/Debugging-with-valgrind).

If the Pull request introduces code that is not covered by the github action (for example coupling with a new library):
  - [ ] Should this use case be added to the github action?
  - [ ] If not, does the specific use case compile and all tests pass (check manually).
#### Scripts and Wiki
- [ ] If a new directory with source files is added, it must be covered by the `script/find_all_source_files.scp` to check the indentation of these files.
- [ ] If this PR introduces a new feature, it must be covered in an example or tutorial and a Wiki article.
#### License
- [ ] The author added a BSD statement to `doc/` (or already has one).

holke committed:
proper abort message when using lfn with unbalanced forest

holke committed:
change leaf parameter name. Add ghost todo comments

holke committed:
adapt iterate face example

holke committed:
fix deleted bracket

holke committed:
use RAII where possible

holke committed:
update iterate face example to ghost face iterate

holke committed:
more const for split_array functions

holke committed:
add ghost check functionality to t8_forest_element_is_leaf

holke committed:
comment and remove empty lines

holke committed:
iterate faces now supports ghosts

holke committed:
debug guards around check used only in debug mode

holke committed:
Merge branch 'feature-more_const' into feature-extend_lfn_to_ghosts

holke committed:
Merge branch 'feature-element_is_ghost' into feature-extend_lfn_to_ghosts

holke committed:
make comment of bin search clearer

holke committed:
start with lfn extension for ghost, have structure done and know what to do now

holke committed:
remove TODO comment

holke committed:
todo comments

holke committed:
Add face iterate callback to leaf face neighbor

holke committed:
is_leaf_or_ghost correct assertion

holke committed:
adaptations in face iterate callback

holke committed:
add ghost support to t8_forest_element_neighbor_eclass

holke committed:
final extensions of lfn to ghost elements

holke committed:
t8_forest_element_neighbor_eclass adaptation to declaration

holke committed:
user data parameter handling for iterate faces

holke committed:
Remove is_balanced parameter from lfn function since no longer required.

holke committed:
improve function documentation

holke committed:
fix abort if incomplete forest

holke committed:
return when no neighbors are found

holke committed:
only compute ghost tree if ghosts exists

holke committed:
skip if case when no neighbors found in search

holke committed:
comments

holke committed:
fix wrongly placed brace

holke committed:
add first test case

holke committed:
only do face neighbors when default quad or hex

holke committed:
fix user data pointer in callback

holke committed:
fix memory counting and handling for computed neighbors

holke committed:
face iterate use is leaf check

holke committed:
adjustments to face neighbor computation

holke committed:
extend lfn test

holke committed:
comments

holke committed:
Face iteration requires view into partial leaf array

holke committed:
store local or ghost tree info with leaf arrays for face iteration

holke committed:
extend forest_element_face_neighbor to ghosts

holke committed:
extend leaf_face_neighbors_iterate to ghosts

holke committed:
new check elements for ancestor function

holke committed:
In face neighbor computation check whether neighbors would exist before searching

holke committed:
more output in test

holke committed:
indent

holke committed:
index checks in face neighbor test

holke committed:
Add own test file for reusable adapt callbacks

holke committed:
add testing of adapted forest to face neighbor test

holke committed:
clean up custom assertion test header

holke committed:
fix pointer arithmetic

holke committed:
extend checks in test for adaptive forest

holke committed:
temporarily skip adapted forest in test

holke committed:
debugging output

holke committed:
proper conversion of pointer output

holke committed:
comment

lukasdreyer committed:
Merge branch 'main' into feature-extend_lfn_to_ghosts

lukasdreyer committed:
leafs -> leaves

lukasdreyer committed:
fix CI

lukasdreyer committed:
adapt test parameter

lukasdreyer committed:
Merge branch 'main' into feature-extend_lfn_to_ghosts

lukasdreyer committed:
indent

lukasdreyer committed:
Merge branch 'feature-extend_lfn_to_ghosts' of github.com:DLR-AMR/t8code into feature-extend_lfn_to_ghosts

sandro-elsweijer committed:
Merge branch 'main' into feature-extend_lfn_to_ghosts

lukasdreyer committed:
Move element_find_owner to public interface

holke committed:
Merge remote-tracking branch 'origin/feature-extend_lfn_to_ghosts' into feature-extend_lfn_to_ghosts

holke committed:
Merge remote-tracking branch 'origin/main' into feature-extend_lfn_to_ghosts

holke committed:
merge updates from fix-forest_leave_face_neigh

holke committed:
update test file from fix-forest_leave_face_neigh branch

holke committed:
add same level face neigh to header

holke committed:
Add TODO comment for adaptive bug and how to solve it

holke committed:
update todo comment

holke committed:
Merge remote-tracking branch 'origin/main' into feature-extend_lfn_to_ghosts

holke committed:
change ifdef to if

holke committed:
Merge branch 'feature-geometry_improved_hashes' into feature-extend_lfn_to_ghosts

holke committed:
Merge remote-tracking branch 'origin/main' into feature-extend_lfn_to_ghosts

holke committed:
update element/leaf function calls

holke committed:
Merge remote-tracking branch 'origin/feature-element_is_ghost' into feature-extend_lfn_to_ghosts

holke committed:
repair merge, is_leaf test reduce cmeshes

holke committed:
improve comments

holke committed:
Change TODO comment to updated strategy - i am very sure that this will work now

holke committed:
update lfn computation to new strategy via same level neighbor. Now only ancestor search missing

holke committed:
indent

holke committed:
Change parameter of iterate_faces to const*const

holke committed:
Merge branch 'feature-ancestor_search' into feature-extend_lfn_to_ghosts

holke committed:
Merge branch 'feature-ancestor_search' into feature-extend_lfn_to_ghosts

holke committed:
remove old elements_are_ancestor

holke committed:
Add new lfn computation via first descendant of same level neighbor

holke committed:
docstring

holke committed:
t8_forest_bin_search_first_descendant_ancenstor fix parameter

holke committed:
restrict search to the descendant of the neighbor element

holke committed:
Merge remote-tracking branch 'origin/main' into feature-extend_lfn_to_ghosts

holke committed:
typo

holke committed:
Add debug output to remove later

holke committed:
rename face neighbor test t8 t8_gtest_leaf_face_neighbors

holke committed:
Set no neighbors found return values in separate function

holke committed:
only compute level if elements exist

holke committed:
Correct off by one error

holke committed:
debug out

holke committed:
Edit docstring for t8_forest_iterate_faces

holke committed:
Add adapt_callback linkage to t8_gtest_element_is_leaf_serial test

holke committed:
Merge remote-tracking branch 'origin/feature-ancestor_search' into feature-ancestor_search

holke committed:
Fix search range computation by computing last element

holke committed:
only reallocate memory if neighbors are found

holke committed:
additional debug check that all leafs to search for are descendants of the element

holke committed:
Simplify the search for the last element in search range

holke committed:
remove old code

holke committed:
indent

holke committed:
fix debug out file format warnings

holke committed:
Add unused parameter flag

holke committed:
add unused flags to adapt callback

holke committed:
typos

holke committed:
typos

holke committed:
add unused flag

Davknapp reviewed (changes requested):
A very small first batch of comments. Need to have a more detailed look into it for a propper review. But thanks for the addition ;)

holke committed:
In leaf face neighbor the succesor now is only computed if it exists.
If no successor exists, we can simply use the last element of the tree's
leaf as end of our search window.

holke committed:
Leaf face neighbor make neighbor_leave pointer to const element and return direct leaf pointers instead of copying elements.

holke committed:
fix wrong class in debug print

holke committed:
indent step6

holke committed:
Add t8_cmesh_get_tree_face_neighbor_eclass function to compute the eclass of a cmesh neighbor tree

Given a local tree or ghost tree and a face this function returns the
eclass of the neighbor tree.
If no neighbor is found it returns T8_ECLASS_INVALID

holke committed:
Adapt t8_forest_element_neighbor_eclass

holke committed:
Use neighbor eclass INVALID argument to exclude computing the face neighbor

holke committed:
extend forest element face neighbor to ghost input

holke committed:
string for invalid eclass

holke committed:
bugfix element_face_neighbor ghost extension

holke committed:
minor edit

holke committed:
Use loop over levels to compute dual face of possible neighbor

holke committed:
Improve element_face_neighbor

holke committed:
Fix t8_forest_element_neighbor_eclass, wrong face passed to cmesh

holke committed:
activate face neighbor computation for all shapes. Still buggy in parallel

holke committed:
drastically simplify t8_forest_element_face_neighbor

holke committed:
skim down lfn neighbor search user data struct

holke committed:
remove old code comment

holke committed:
remove unneeded eclass var

holke committed:
Merge remote-tracking branch 'origin/main' into feature-extend_lfn_to_ghosts

holke committed:
Merge remote-tracking branch 'origin/fix-1951_half_face_neighbor_order' into feature-extend_lfn_to_ghosts

holke committed:
use ECLASS_INVALID in half neighbor test

holke committed:
correct typos

holke committed:
Merge branch 'fix-vtk_dontcomputeboundsforemptyproc' into feature-extend_lfn_to_ghosts

holke commented on src/t8_cmesh.h:
These changes are already in #1806

holke commented on src/t8_data/t8_containers.cxx:
These changes are adressed in #1301

holke committed:
Merge branch 'feature-ancestor_face' into feature-extend_lfn_to_ghosts

holke committed:
use new ancestor face function

holke commented on src/t8_forest/t8_forest_balance.cxx:
```diff
@@ -81,34 +81,39 @@ t8_forest_balance_adapt (t8_forest_t forest, t8_forest_t forest_from, const t8_l
-                                                             neigh_class, iface, num_half_neighbors, NULL);
-      if (neighbor_tree >= 0) {
-        /* The face neighbors do exist, check for each one, whether it has
+      if (neigh_class != T8_ECLASS_INVALID) {
```
I only added this line and the "if (neighbor_tree >= 0) {" below.
The rest is just indentation.

holke commented on example/forest/t8_test_face_iterate.cxx:
I am not sure why i edited this and what i did here.
Its too long ago.

holke commented on src/t8_forest/t8_forest.cxx:
```diff
@@ -1414,119 +1417,82 @@ t8_forest_copy_trees (t8_forest_t forest, t8_forest_t from, int copy_elements)
   }
 }
 
```
I simplified the t8_forest_element_neighbor_eclass  function.
It works with elements and ghosts now.

holke committed:
more const params

holke committed:
Merge branch 'feature-face_iterate_const' into feature-extend_lfn_to_ghosts

holke committed:
const function callback params

holke commented on src/t8_forest/t8_forest.cxx:
```diff
@@ -1634,253 +1631,454 @@ t8_forest_leaf_face_orientation (t8_forest_t forest, const t8_locidx_t ltreeid,
+static int
+t8_forest_leaf_face_neighbors_iterate (t8_forest_t forest, t8_locidx_t ltreeid, const t8_element_t *element, int face,
+                                       int is_leaf, [[maybe_unused]] const t8_element_array_t *const leaf_elements,
+                                       t8_locidx_t tree_leaf_index, void *user_data)
```
Done

holke commented on src/t8_forest/t8_forest.cxx:
```diff
@@ -1634,253 +1631,454 @@ t8_forest_leaf_face_orientation (t8_forest_t forest, const t8_locidx_t ltreeid,
+  // compute its id as a ghost tree ( 0 <= id < num_ghost_trees)
+  const bool is_ghost_tree = !t8_forest_tree_is_local (forest, ltreeid);
+  const t8_locidx_t adjusted_tree_id = !is_ghost_tree ? ltreeid : ltreeid - t8_forest_get_num_local_trees (forest);
+  T8_ASSERT (t8_forest_element_is_leaf_or_ghost (forest, element, adjusted_tree_id, is_ghost_tree));
```
I checked again and did not find a suitable function.
There is t8_forest_get_local_or_ghost_id but this returns an id >= num_local_trees for the ghosts, which is not what we need here.

holke commented on src/t8_forest/t8_forest.cxx:
```diff
@@ -1961,36 +2131,36 @@ t8_forest_tree_is_local (const t8_forest_t forest, const t8_locidx_t local_tree)
 int
```
Changes until line 2165 are adressed in #1301 (add is_ghost is_leaf_or_ghost)

holke commented on src/t8_forest/t8_forest_ghost.h:
See #1301

holke commented on src/t8_forest/t8_forest_iterate.cxx:
I extended the iteration to ghost elements.

holke commented on src/t8_schemes/t8_scheme.hxx:
Adressed in #1714 (element_is_ancestor) and #1958 (element_face_get_ancestor_face)

holke commented on src/t8_schemes/t8_standalone/t8_standalone_implementation.hxx:
Adressed in #1714

holke committed:
remove unneeded debug code that skipped cmeshes in testing

holke commented on test/t8_gtest_adapt_callbacks.cxx:
Added an extra file to reuse adaptation callbacks in tests.
This could be an extra PR as well, if you think so.

holke commented on test/t8_schemes/t8_gtest_is_ancestor.cxx:
See #1714

holke commented on test/t8_schemes/t8_gtest_nca.cxx:
See #1714

holke committed:
bugfix order of parameters in function call

holke committed:
remove test code that was left over from debugging and did nothing

holke committed:
protect debugging code in debugging macros

holke committed:
rename parameter to satisfy typo script

holke committed:
element maybe unused in iterate callback

holke committed:
Its now possible to call leaf face neighbor if forest does not have ghosts

holke committed:
Improve docu

holke committed:
maybe unused params for adapt callback

holke committed:
Merge branch 'main' into feature-extend_lfn_to_ghosts

holke committed:
Merge remote-tracking branch 'origin/main' into feature-extend_lfn_to_ghosts

holke committed:
Merge remote-tracking branch 'origin/feature-extend_lfn_to_ghosts' into feature-extend_lfn_to_ghosts

holke committed:
change include paths

Davknapp committed:
Merge branch 'main' into feature-extend_lfn_to_ghosts

holke committed:
Merge remote-tracking branch 'origin/feature-ancestor_face' into feature-extend_lfn_to_ghosts

holke committed:
Merge remote-tracking branch 'origin/feature-ancestor_search' into feature-extend_lfn_to_ghosts

holke committed:
Fix linker issue since helper file was not compiled for test

holke committed:
Fix memory error, neighbor element was not destroyed

holke committed:
Fix documentation

holke committed:
Adapt mesh handle to new face neighbor interface

holke committed:
Merge remote-tracking branch 'origin/main' into feature-extend_lfn_to_ghosts

holke committed:
indent

holke committed:
Activate lfn test. No idea why it was not active before

holke committed:
Add specific lfn test case with 2 square twisted mesh file

holke committed:
Add NEWS.md text

holke commented:
> Could the changed freeing policy of `t8_forest_leaf_face_neighbors` be documented somewhere?

I added a text in the NEWS.md file.

holke committed:
Merge remote-tracking branch 'origin/main' into feature-extend_lfn_to_ghosts

holke committed:
Merge branch 'main' into feature-extend_lfn_to_ghosts

holke committed:
Document adapt callback

holke committed:
fix merge

holke committed:
remove blank space

holke committed:
adapt path to new testfile path

holke committed:
remove lambda left over from merge

holke committed:
reduce test levels

holke committed:
change how we get the element

Davknapp committed:
Merge branch 'main' into feature-extend_lfn_to_ghosts

holke committed:
revert change how we get element. Need to access array since it could be tree or ghost

holke committed:
Merge remote-tracking branch 'origin/feature-extend_lfn_to_ghosts' into feature-extend_lfn_to_ghosts

holke committed:
Try fixing mesh file path

holke committed:
Merge remote-tracking branch 'origin/main' into feature-extend_lfn_to_ghosts

holke committed:
Add more tests in msh file example and add comment

Davknapp committed:
Merge branch 'main' into feature-extend_lfn_to_ghosts

Davknapp reviewed (commented):
A batch of review comments.

holke committed:
Merge with main

holke committed:
indent

holke committed:
Merge with main

holke committed:
Merge with main

holke committed:
Merge with main

holke committed:
Merge branch 'main' into feature-extend_lfn_to_ghosts

holke committed:
typo

holke committed:
use new test data dir to locate test file

holke committed:
Update date in copyright info

Co-authored-by: David Knapp <david.knapp@dlr.de>

holke commented on test/t8_forest/t8_gtest_leaf_face_neighbors.cxx:
```diff
@@ -0,0 +1,546 @@
+    }
+    const t8_scheme *scheme = create_from_scheme_id (scheme_id);
+    const int level = 0;
+    const int adapt_levels = 2;
```
I would keep it.
We start with a uniform level L and from that we adapt R levels to reach the final level M.
To make the code more readable and changable it feels useful to have these define as different variables.

I will change the name "level" to "uniform_level" in order to better capture the meaning of the variable.

holke committed:
user better varname uniform_level in test

holke committed:
change ?: operator logic to make code more readable

Co-authored-by: David Knapp <david.knapp@dlr.de>

holke committed:
check element index if it is a ghost - part 1

holke committed:
Merge remote-tracking branch 'origin/feature-extend_lfn_to_ghosts' into feature-extend_lfn_to_ghosts

holke committed:
fix wrong index

holke committed:
unify element index verification in neighbor and neighbor neighbor case.

holke committed:
remove old fflush

Davknapp commented on src/t8_forest/t8_forest_general.h:
```diff
@@ -599,32 +599,28 @@ t8_forest_leaf_face_orientation (t8_forest_t forest, const t8_locidx_t ltreeid,
- * \note Currently \a forest must be balanced.
  * \note \a forest must be committed before calling this function.
- *
+ * \note If \a forest does not have a ghost layer then leaf elements at the process boundaries have 0 neighbors. (The function output for leaf elements then depends on the parallel partition.)
```
```suggestion
 * \note If \a forest does not have a ghost layer then leaf elements at the process boundaries have 0 neighbors along the boundary face. (The function output for leaf elements then depends on the parallel partition.)
```

holke committed:
Typo

holke committed:
rearrange ternary operator logic

Co-authored-by: David Knapp <david.knapp@dlr.de>

holke committed:
More const.

Co-authored-by: David Knapp <david.knapp@dlr.de>

holke committed:
reorder ternary operator

Co-authored-by: David Knapp <david.knapp@dlr.de>

holke committed:
Apply suggestion from @Davknapp

Co-authored-by: David Knapp <david.knapp@dlr.de>

holke committed:
Add a comment on T8_FREE usage in test

holke committed:
indent

holke committed:
Merge remote-tracking branch 'origin/feature-extend_lfn_to_ghosts' into feature-extend_lfn_to_ghosts

holke committed:
Specify neighbor face in comment

Co-authored-by: David Knapp <david.knapp@dlr.de>

holke committed:
Add a comment referring a TODO to an issue

holke committed:
Change variable name for better readability

Co-authored-by: David Knapp <david.knapp@dlr.de>

holke committed:
Change ternary operator logic

Co-authored-by: David Knapp <david.knapp@dlr.de>

holke committed:
Change if else evaluation logic

Co-authored-by: David Knapp <david.knapp@dlr.de>

holke committed:
Add comment to t8_forest_leaf_face_neighbors_set_no_neighbor_return_value

holke committed:
remove done TODO comment

holke committed:
Merge remote-tracking branch 'origin/feature-extend_lfn_to_ghosts' into feature-extend_lfn_to_ghosts

holke committed:
Change TODO comment and need to investigate

holke committed:
indent

holke commented on src/t8_forest/t8_forest.cxx:
```diff
@@ -1634,253 +1631,454 @@ t8_forest_leaf_face_orientation (t8_forest_t forest, const t8_locidx_t ltreeid,
+    *gneigh_tree = -1;
+  }
+}
+
+void
+t8_forest_leaf_face_neighbors_ext (t8_forest_t forest, t8_locidx_t ltreeid, const t8_element_t *leaf_or_ghost,
+                                   t8_element_t **pneighbor_leaves[], int face, int *dual_faces[], int *num_neighbors,
+                                   t8_locidx_t **pelement_indices, t8_eclass_t *pneigh_eclass, t8_gloidx_t *gneigh_tree,
+                                   int *orientation)
```
Should be done

holke committed:
Change nullptr to NULL

Co-authored-by: David Knapp <david.knapp@dlr.de>

holke committed:
Correct copyright year

Co-authored-by: David Knapp <david.knapp@dlr.de>

holke committed:
Change documentation from "is adapted" to "uniform"

Co-authored-by: David Knapp <david.knapp@dlr.de>

holke committed:
clarify face_iterate todos

holke commented on src/t8_forest/t8_forest_iterate.h:
```diff
@@ -150,6 +153,10 @@ T8_EXTERN_C_BEGIN ();
 void
 t8_forest_split_array (const t8_element_t *element, const t8_element_array_t *leaf_elements, size_t *offsets);
 
+/* TODO: comment */
```
I added an issue for the first TODO.

The latter two are more dev comments. Should we C++- restructure this function they might be useful.
But we might decide to not do them at all,
I did not feel like an issue would be necessary here. i added "maybe" to the comment.

holke commented on test/t8_forest/t8_gtest_leaf_face_neighbors.cxx:
```diff
@@ -0,0 +1,546 @@
+
+            // clean-up neighbor's neighbors
+            if (neigh_num_neighbors > 0) {
+              T8_FREE (neigh_neighbor_leaves);
```
Done

holke reviewed (commented):
Thank you for the review @Davknapp!
This was a massive one.

I am done with adressing all your comments.

holke committed:
Merge with main

holke committed:
Merge with main again

holke committed:
Merge branch 'main' into feature-extend_lfn_to_ghosts

holke commented:
> > I am waiting to resolve the merge conflicts until i have included your review comments. Merging this with main is always annoying and time consuming to get it right, so i'd rather do it once at the end.
> 
> Ok, then assign me again as soon as you have addressed my comments.

I adressed all your comments and merged with main.

holke committed:
indent

holke committed:
Merge remote-tracking branch 'origin/feature-extend_lfn_to_ghosts' into feature-extend_lfn_to_ghosts

Davknapp reviewed (changes requested):
Maybe you've missed it, but there is one (small) todo left open. Please address it!

holke committed:
Apply suggestions from code review - move variable declaration into for-loop

Co-authored-by: David Knapp <david.knapp@dlr.de>

holke committed:
Merge remote-tracking branch 'origin/main' into feature-extend_lfn_to_ghosts

holke committed:
Merge remote-tracking branch 'origin/feature-extend_lfn_to_ghosts' into feature-extend_lfn_to_ghosts
